cmd/zoekt-webserver/main.go

```diff
@@ -504,3 +510,70 @@ var (
 		Help: "The total number of errors from zoekt watchdog.",
 	})
 )
+
+func mustRegisterMemoryMapMetrics(l sglog.Logger) {
+	if runtime.GOOS != "linux" {
+		l.Debug(
+			"skipping registration: must be running on a linux-based operating system",
+			sglog.String("current_operating_system", runtime.GOOS),
+			sglog.String("desired_operating_system", "linux"))
+
+		return
+	}
+
+	fs, err := procfs.NewDefaultFS()
+	if err != nil {
+		panic(fmt.Sprintf("registering memory map metrics: failed to initialize proc FS: %s", err))
+	}
+
+	prometheus.MustRegister(prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "zoekt_webserver_max_memory_map_count",
+		Help: "Upper limit on amount of memory mapped regions a process may have.",
+	}, func() float64 {
+		vm, err := fs.VM()
+		if err != nil {
+			l.Debug(
+				"failed to read VM statistics",
+				sglog.String("path", path.Join(procfs.DefaultMountPoint, "sys", "vm")),
+				sglog.String("error", err.Error()),
+			)
+
+			return 0
+		}
+
+		if vm.MaxMapCount == nil {
+			return 0
+		}
+
+		return float64(*vm.MaxMapCount)
+	}))
+
+	prometheus.MustRegister(prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "zoekt_webserver_current_memory_map_count",
+		Help: "Amount of memory mapped regions this process is currently using.",
+	}, func() float64 {
+		self, err := fs.Self()
+		if err != nil {
+			l.Debug(
+				"failed to read process statistics",
+				sglog.String("path", path.Join(procfs.DefaultMountPoint, "self")),
+				sglog.String("error", err.Error()),
+			)
+
+			return 0
+		}
+
+		procMaps, err := self.ProcMaps()
```

Comment:
this looks very expensive for us. We can have _lots_ of mmaped files. Is there not another way to get this count? Alternatively I'd rather just use the result of readdir on the relevant directory than parse every single file.

For example there is already a FileDescriptorsLen method to avoid the parsing of each one, but this doesn't exist for mmap.

Reply:
`/proc/self/maps` is a single file, not a directory. See the man page:

> /proc/[pid]/maps
>               A file containing the currently mapped memory regions and
>               their access permissions.  See [mmap(2)](https://man7.org/linux/man-pages/man2/mmap.2.html) for some further
>               information about memory mappings.
> 
>               Permission to access this file is governed by a ptrace
>               access mode PTRACE_MODE_READ_FSCREDS check; see [ptrace(2)](https://man7.org/linux/man-pages/man2/ptrace.2.html).
> 
>               The format of the file is:
> 
>                   address           perms offset  dev   inode       pathname
>                   00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/dbus-daemon
>                   00651000-00652000 r--p 00051000 08:02 173521      /usr/bin/dbus-daemon
>                   00652000-00655000 rw-p 00052000 08:02 173521      /usr/bin/dbus-daemon


I don't see another method to do this while staying within the procfs package. Another alternative is to read /proc/self/maps directly and count the number of newlines, but I don't think its worth loosing the consistency of going through procfs for anything. The [implementation](https://sourcegraph.com/github.com/prometheus/procfs@f5d82fa2c83349c3e3ea8587cd4f5a5983b99a9e/-/blob/proc_maps.go?L189-211) for this just seems to do some string parsing (while, which its unnecessary, doesn't seem that expensive for something that isn't in the hot path of any other functionality). 

Let me know if you disagree though

Reply:
aah ok, nevermind then this seems good. I haven't eyeballed pprof from prod ni a while, when this change goes out lets take a look to see if this metric pops up :)